components/src/core/app-bar/app-bar.component.ts

```diff
@@ -0,0 +1,173 @@
+
+    // The thing that scrolls, we listen to this.
+    @Input() scrollContainerElement: Element;
+    @Input() scrollContainerClassName: { name: string; index: number };
+    @Input() scrollContainerId: string;
+    scrollEl;
+
+    isWindow = false;
+
+    @Output() onReachedExpandedHeight: EventEmitter<void> = new EventEmitter();
+    @Output() onReachedCollapsedHeight: EventEmitter<void> = new EventEmitter();
+
+    currentHeight: number;
+    scrollDistance: number;
+    useDefaultCollapsedHeight = false;
+    toolbar: HTMLElement;
+
+    constructor(private readonly _ref: ChangeDetectorRef) {}
+
+    ngOnInit(): void {
+        this.currentHeight = this.expandedHeight;
+    }
+
+    ngOnChanges(changes: SimpleChanges): void {
+        this.useDefaultCollapsedHeight = !this.collapsedHeight || this.collapsedHeight === 0;
+        if (changes.collapsedHeight || changes.expandedHeighted) {
+            this.expandedHeight = Number(this.expandedHeight);
+            this.collapsedHeight = Number(this.collapsedHeight);
+        }
+        if (changes.mode) {
+            this._resizeOnModeChange();
+        }
+    }
+
+    ngAfterViewInit(): void {
+        this._setScrollEl();
+        this.toolbar = this.bar.nativeElement;
+        this._resizeOnModeChange();
+        fromEvent(this.scrollEl, 'scroll')
+            .pipe(throttle(() => interval(10)))
+            .subscribe((event: Event) => {
+                this._resizeEl(event);
+            });
+
+        fromEvent(window, 'resize').subscribe((event: Event) => {
+            this._resizeEl(event);
+        });
+    }
+
+    private _resizeOnModeChange(): void {
+        if (this.mode !== 'collapsible') {
+            return this._handleLockedModes();
+        } else if (this.scrollEl) {
+            this._resizeEl();
+        }
+    }
+
+    private _handleLockedModes(): void {
+        if (this.mode === 'lock-collapsed') {
+            this.currentHeight = this.useDefaultCollapsedHeight
+                ? this._calcDefaultCollapsedHeight()
+                : this.collapsedHeight;
+            this.toolbar.style.marginTop = `0px`;
+            this._ref.detectChanges();
+            return;
+        }
+        if (this.mode === 'lock-expanded') {
+            this.currentHeight = this.expandedHeight;
+            this.toolbar.style.marginTop = `0px`;
+            this._ref.detectChanges();
+            return;
+        }
+    }
+
+    private _resizeEl(event?: Event): void {
+        if (this.mode !== 'collapsible') {
+            return this._handleLockedModes();
+        }
+
+        const collapsedHeight = this.useDefaultCollapsedHeight
+            ? this._calcDefaultCollapsedHeight()
+            : this.collapsedHeight;
+        this.collapsedHeight = collapsedHeight;
+
+        const el = this.scrollEl;
+        this.scrollDistance = this.isWindow ? document.scrollingElement.scrollTop : el.scrollTop;
+        if (this.scrollDistance === 0) {
```

Comment:
should we be checking against the scroll distance or against the element height? The event says "xxx_height_reached"

Reply:
I think whenever the scroll distance is 0, then the element height should be maxed out. They should mean the same thing, right?

Reply:
Yes, the end result right now would be the same...I was thinking if perhaps in the future there's not a 1-1 mapping of scroll distance to height...like if we go back and do the snapping stuff.